tools/querytee/proxy_endpoint.go

```diff
@@ -85,20 +86,41 @@ func (p *ProxyEndpoint) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	p.metrics.responsesTotal.WithLabelValues(downstreamRes.backend.Name(), r.Method, p.route.RouteName).Inc()
 }
 
-func (p *ProxyEndpoint) selectBackends() []ProxyBackendInterface {
-	if len(p.backends) == 1 || p.secondaryBackendRequestProportion == 1.0 {
+func (p *ProxyEndpoint) selectBackends(r *http.Request) []ProxyBackendInterface {
+	if len(p.backends) == 1 {
 		return p.backends
 	}
 
-	if p.secondaryBackendRequestProportion == 0.0 {
-		return []ProxyBackendInterface{p.preferredBackend}
+	minQueryTime := extractMinTimeFromRequest(r)
+
+	selected := []ProxyBackendInterface{}
+
+	if p.preferredBackend != nil {
+		selected = append(selected, p.preferredBackend)
 	}
 
-	if rand.Float64() > p.secondaryBackendRequestProportion {
-		return []ProxyBackendInterface{p.preferredBackend}
+	for _, backend := range p.backends {
+		if backend.Preferred() {
+			continue
+		}
+
+		if !backend.ShouldHandleQuery(minQueryTime) {
+			continue
+		}
+
+		var proportion float64
+		if backend.HasConfiguredProportion() {
+			proportion = backend.RequestProportion()
+		} else {
+			proportion = p.secondaryBackendRequestProportion
+		}
+
+		if proportion == DefaultRequestProportion || rand.Float64() <= proportion {
```

Comment:
Similarly, I think this first check `proportion == DefaultRequestProportion` can catch us by surprise if we change `proportion == DefaultRequestProportion`. Doesn't it work the same way without it?